README.md
# PS2Mouse
library for connecting Synaptics TouchPad to arduino

[![PlatformIO CI](https://github.com/RootShell-coder/PS2Mouse/actions/workflows/main.yml/badge.svg)](https://github.com/RootShell-coder/PS2Mouse/actions/workflows/main.yml)

## Интерфейс PS/2
Протокол PS/2 включает два сигнальных провода, а также питание +5v и землю. Сигнальные провода, CLK и DATA, представляют собой двунаправленные сигналы с открытым стоком. Они удерживаются на высоком уровне (+5v) подтягивающим резистором 5–10 кОм на хосте. Хост или сенсорная панель могут в любой момент снизить уровень CLK или DATA. Когда порт бездействует, оба сигнальных провода находятся на высоком уровне. Хост может заблокировать устройство в любое время, удерживая CLK на низком уровне. Линии CLK и DATA должны иметь общую емкость не более 500pf, чтобы подтягивающий резистор 5–10 кОм мог вывести их на высокий уровень напряжения за разумное время. 

## Передача байт
Стандартная мышь PS/2 отправляет информацию о движении (и кнопках) на хост, используя следующий 3-байтовый пакет
<table>
    <tr>
        <td></td>
        <td>Bit 7</td>
        <td>Bit 6</td>
        <td>Bit 5</td>
        <td>Bit 4</td>
        <td>Bit 3</td>
        <td>Bit 2</td>
        <td>Bit 1</td>
        <td>Bit 0</td>
    </tr>
    <tr>
        <td>Byte 1</td>
        <td>Y overflow</td>
        <td>X overflow</td>
        <td>Y sign bit</td>
        <td>X sign bit</td>
        <td>Always 1</td>
        <td>Middle Btn</td>
        <td>Right Btn </td>
        <td>Left Btn</td>
    </tr>
    <tr>
        <td>Byte 2</td>
        <td colspan="8" align="center">X Movement 7...0</td>
    </tr>
    <tr>
        <td>Byte 3</td>
        <td colspan="8" align="center">Y Movement 7...0</td>
    </tr>
</table>

Счетчики движения представляют собой 9-битные целые числа с дополнением до 2, где самый старший бит появляется как знаковый бит в байте 1 пакета данных движения. Эти счетчики обновляются, когда мышь считывает ввод и обнаруживает, что произошло движение. Их значение - это количество перемещений, которое произошло с момента отправки последнего пакета данных о перемещении на хост (т.е. После того, как пакет отправлен на хост, счетчики перемещений сбрасываются). Диапазон значений, который может быть выражен с помощью счетчиков движения от -255 до +255. Если этот диапазон превышен, устанавливается соответствующий бит переполнения, и счетчик не увеличивается/уменьшается до тех пор, пока он не будет сброшен. Сброс происходит всякий раз, когда пакет данных движения успешно отправляется на хост. Они также сбрасываются после того, как мышь получает любую команду от хоста, кроме команды "Отправить повторно" (0xFE). 

Каждый байт, передаваемый между устройством и хостом, включает в себя стартовый бит (логический 0), восемь бит данных (сначала младший бит), бит четности (нечетная четность) и стоповый бит (логическая 1).
![11 bit](https://github.com/RootShell-coder/PS2Mouse/blob/master/add/Byte%20transmission.png "11 bit")
 _Нечетность означает, что восемь битов данных и бит четности вместе содержат нечетное количество единиц._
 
Каждый байт команды создает по крайней мере один байт ответа от сенсорной панели. Для каждой команды, кроме команды "Отправить повторно" (0xFE), ответ всегда начинается с байта "Подтверждение" ACK (0xFA).

## Инициализация
Пример инициализации microsoft scrolling mouse от [Adam Chapweske](https://isdaman.com/alsos/hardware/mouse/ps2interface.htm)
```bash
Power-on Reset               : Включение питания             
Mouse: AA  Self-test passed  : Мышь отправляет 0xAA Успешная самодиагностика
Mouse: 00  Mouse ID          : Мышь отправляет свой DI 0x00 
Host:  FF  Reset command     : Хост отправляет команду перезагрузки 0xFF
Mouse: FA  Acknowledge       : ACK подтверждение от мыши о том что команда принята 0xFA
Mouse: AA  Self-test passed  : При инициализации 3 раза хост отправляет команду 0xFF (перезагрузка)
Mouse: 00  Mouse ID          
Host:  FF  Reset command     
Mouse: FA  Acknowledge       
Mouse: AA  Self-test passed  
Mouse: 00  Mouse ID          
Host:  FF  Reset command     
Mouse: FA  Acknowledge       
Mouse: AA  Self-test passed  
Mouse: 00  Mouse ID           
Host:  F3  Set Sample Rate    
Mouse: FA  Acknowledge       
Host:  C8  decimal 200       
Mouse: FA  Acknowledge       
Host:  F3  Set Sample Rate
Mouse: FA  Acknowledge       
Host:  64  decimal 100       
Mouse: FA  Acknowledge       
Host:  F3  Set Sample Rate   
Mouse: FA  Acknowledge       
Host:  50  decimal 80        
Mouse: FA  Acknowledge       
Host:  F2  Read Device Type  
Mouse: FA  Acknowledge       
Mouse: 00  Mouse ID          
Host:  F3  Set Sample Rate 
Mouse: FA  Acknowledge
Host:  0A  decimal 10
Mouse: FA  Acknowledge
Host:  F2  Read Device Type
Mouse: FA  Acknowledge
Mouse: 00  Mouse ID
Host:  E8  Set resolution
Mouse: FA  Acknowledge
Host:  03  8 Counts/mm
Mouse: FA  Acknowledge
Host:  E6  Set Scaling 1:1
Mouse: FA  Acknowledge
Host:  F3  Set Sample Rate
Mouse: FA  Acknowledge
Host:  28  decimal 40
Mouse: FA  Acknowledge
Host:  F4  Enable
Mouse: FA  Acknowledge
Initialization complete...
```

### Stream Mode:
В потоковом режиме мышь отправляет данные о движении, когда обнаруживает движение или изменение состояния одной или нескольких кнопок мыши. Максимальная частота, с которой может происходить этот отчет, известна как частота дискретизации. Этот параметр находится в диапазоне от 10 выборок/сек до 200 выборок/сек. Его значение по умолчанию составляет 100 выборок в секунду, и хост может изменить это значение с помощью команды "Установить частоту дискретизации" (0xF3). 
Потоковый режим - это режим работы по умолчанию и этот режим реализован в библиотеке PS2Mouse.

### Remote Mode:
В этом режиме мышь считывает свои входы и обновляет свои счетчики/флаги с текущей частотой дискретизации, но она только уведомляет хост о движении (и изменении состояния кнопки), когда эта информация запрашивается хостом. Хост делает это, выдавая команду "Прочитать данные" (0xEB). После получения этой команды мышь отправит пакет данных о движении и сбросит свои счетчики движения. 

### Relative mode
Относительный режим является режимом по умолчанию для большинства сенсорных панелей Synaptics. При подаче питания сенсорная панель идентифицируется для главного компьютера как обычная мышь. Это позволяет использовать сенсорную панель со стандартными драйверами мыши. Этот режим, совместимый с мышью, называется относительным режимом, потому что действия пальца передаются хосту в виде относительных движений мыши по поверхности сенсора. TouchPad сообщает об этом относительном движении хосту в пакетах, совместимых с мышью. Каждый пакет сообщает о количестве движения в направлениях X (по горизонтали) и Y (по вертикали), которое произошло с момента предыдущего пакета. Эти количества движения называются дельтами и обозначаются "∆X" и "∆Y". Движение слева направо сообщается с положительным значением ∆X, а движение справа налево - с отрицательным значением ∆X. О восходящем движении сообщается положительное значение ∆Y, а о нисходящем движении - отрицательное значение ∆Y. Пакет также сообщает информацию о кнопках «мыши». Поскольку относительный пакет разработан для совместимости с существующим протоколом мыши, точное содержимое относительного пакета варьируется от одного протокола к другому. По умолчанию используется двухкнопочная мышь.

### Absolute mode
Сенсорные панели Synaptics поддерживают абсолютный режим работы, при котором сенсорная панель передает расширенный пакет, который сообщает об абсолютном положении пальца на панели (X, Y), давлении пальца (Z) и другой информации, включая состояние кнопки. Cенсорная панель Synaptics начинает отправлять пакеты, когда Z равно 8 или больше. Сенсорная панель также начинает отправлять пакеты при каждом нажатии или отпускании любой кнопки. Как только сенсорная панель начинает передачу, она продолжает отправлять пакеты в течение одной секунды после того, как Z упадет ниже 8 и кнопки перестанут нажиматься. TouchPad делает это отчасти для того, чтобы программное обеспечение хоста могло использовать поток пакетов в качестве временной базы для декодирования жестов.


## Absolute packet format
Когда включен абсолютный режим, каждый отчет о движении состоит из шести байт. Эти байты кодируют абсолютное положение пальца по осям X, Y на поверхности датчика, а также значение Z (давление) и различные другие измерения и биты состояния.
![6 Byte](https://github.com/RootShell-coder/PS2Mouse/blob/master/add/6byte%20Wmode.png "6 Byte")
Для Synaptics touchPad этот режим включается при инициализации путем отправки последовательности команд. Например, чтобы установить байт режим на 0xC1 (абсолютный режим, высокая скорость передачи пакетов, включен режим Wmode), можно использовать последовательность команд, `0xE8,   0x03,   0xE8,   0x00,   0xE8,   0x00,   0xE8,   0x01,   0xF3,   0x14` где аргумент 0xC1 кодируется следующим образом: `(0x03 × 64) + (0x00 × 16) + (0x00 × 4) + 0x01  =  $C1` 


<table>
    <tr>
        <td></td>
        <td>bit 7<br></td>
        <td>bit 6<br></td>
        <td>bit 5<br></td>
        <td>bit 4<br></td>
        <td>bit 3</td>
        <td>bit 2</td>
        <td>bit 1</td>
        <td>bit 0<br></td>
    </tr>
    <tr>
        <td>byte 1<br></td>
        <td>1</td>
        <td>0</td>
        <td colspan="2">w value 3..2 <br></td>
        <td>0</td>
        <td>w val 1</td>
        <td>right</td>
        <td>left <br></td>
    </tr>
    <tr>
        <td>byte 2<br></td>
        <td colspan="4" align="center">y position 11..8</td>
        <td colspan="4" align="center">x position 11..8<br></td>
    </tr>
    <tr>
        <td>byte 3<br></td>
        <td colspan="8">z pressure 7..0</td>
    </tr>
    <tr>
        <td>byte 4<br></td>
        <td>1</td>
        <td>1</td>
        <td>y pos 12</td>
        <td>x pos 12</td>
        <td>0</td>
        <td>w val 0<br></td>
        <td>r/d</td>
        <td>l/u</td>
    </tr>
    <tr>
        <td>byte 5<br></td>
        <td colspan="8" align="center">x position 7..0</td>
    </tr>
    <tr>
        <td>byte 6<br></td>
        <td colspan="8" align="center">y position 7..0 <br></td>
    </tr>
</table>

_absolute X/Y/Z/W motion packet (Wmode = 1)_

На типичных сенсорных панелях бит L/U идентичен биту левой кнопки, а бит R/D идентичен биту Right. На пэдах MultiSwitch с установленным битом возможностей capFourButtons и включенным Wmode биты L/U и R/D также сообщают о положениях кнопок Up и Down соответственно. Бит L/U сообщает о логическом исключающем ИЛИ для состояний кнопок "Влево" и "Вверх". С другой стороны, L/U - это то же самое, что и левый бит, если только не нажата кнопка Up, и в этом случае L/U является дополнением к левому биту. Бит R/D аналогичным образом выполняет XOR кнопок вправо и вниз. Это кодирование гарантирует, что пакет будет обратно совместим (и устойчив к вмешательству со стороны "слишком умных" контроллеров клавиатуры) всякий раз, когда не нажимаются кнопки "Вверх" и "Вниз". Некоторые сенсорные панели MultiSwitch используют другой формат, в котором можно разместить до 11 кнопок. Пакет абсолютного режима, как и пакет относительного режима, содержит несколько бит, которые сообщают о состоянии кнопок. Важно отметить, что в абсолютном режиме физические кнопки сообщаются отдельно от жестов касания и скольжения. В относительном режиме, совместимом с мышью, жесты и кнопки смешиваются, и хост не может их различить. Сенсорные панели Synaptics поддерживают дополнительное значение в пакете Absolute, называемое W mode. Значение W сообщается только тогда, когда хост включает специальный режим W. Значение W предоставляет дополнительную информацию о характере контакта с датчиком. Хост может использовать W, чтобы различать нормальные пальцы, случайный контакт ладони и несколько пальцев. 

## Подключение
![Synaptics T1004B](https://github.com/RootShell-coder/PS2Mouse/blob/master/add/Synaptics%20T1004B.png "Synaptics T1004B")
![TM4PUF1372 WH511-095](https://github.com/RootShell-coder/PS2Mouse/blob/master/add/TM4PUF1372%20WH511-095.png "TM4PUF1372 WH511-095")

Подключение к nodemcu V3 (esp8266)
| touchpad   |       |  nodemcu V3    |
|----|-------|------|
|T22 | (3.3v) |  3V |
|T10 | (SCL) | GPIO5 (D1)|
|T11 |(SDA/DATA) | GPIO4 (D2)|
|T23 |(GND)  |GROUND |

При успешной инициализации Synaptics touchpad в консоль будет выведен текст

`TouchPad: Synaptics, id: 1000111, minor rev.: 1001, major: 10101, state: init completed`

При этом активируется расширенный, 6 байтовый режим обмена данными. absolute X/Y/Z/W motion packet (Wmode = 1) и как только будут получены данные от touthpad они через функцию прерывания handleInterrupt запишутся в массив raw и в loop выведутся сырые данные.
```C++
#include <ESP8266WiFi.h>
#include <PS2Mouse.h>


PS2Mouse mouse(SCL, SDA);
void IRAM_ATTR handleInterrupt();


void setup() {
  Serial.begin(115200);
  mouse.begin();
  attachInterrupt(digitalPinToInterrupt(SCL), handleInterrupt, FALLING);

}

byte raw[6];

void handleInterrupt() {
  noInterrupts();
    for(int i = 0; i < 6; i++){
      raw[i] = mouse.get(true);
    }
  interrupts();
}

void loop() {
  if(raw[0] != 0 && raw[0] != 0x80){
        Serial.print(raw[2],BIN);
        Serial.print("\t");
        Serial.print(raw[3],BIN);
        Serial.print("\t");
        Serial.print(raw[4],BIN);
        Serial.print("\t");
        Serial.println(raw[5],BIN);
    }
}
```
Если по какой то причине не удается получить абсолютный режим инициализируется 3х байтный Relative mode и массив raw нужно использовать длинной 3 при этом пересчитав циклы для этого массива.

Для получения точнных координат [examples](https://github.com/RootShell-coder/PS2Mouse/blob/master/examples/absoluteXYZ.ino)
```c++
void loop() {
    if(raw[0] != 0 && raw[0] != 0x80){
      mouse.getAbsoluteAxis(raw, x, y);
      Serial.print("Z pressure ");
      Serial.print(raw[2]);
      Serial.print("\t");
      Serial.print("X position ");
      Serial.print(x);
      Serial.print("\t");
      Serial.print("Y position ");
      Serial.print(y);
      Serial.println();
    }
```


CHANGELOG.md
# Changelog
## [0.0.4] - 2021-12-03
non-code fixes
## [0.0.3] - 2021-12-02
Work tested on Synaptics touchpad TM4PUF1372 WH511-095 and nodemsuv3 board


example/main.cpp
#include <ESP8266WiFi.h>
#include <PS2Mouse.h>

PS2Mouse mouse(SCL, SDA);
void IRAM_ATTR handleInterrupt();

void setup() {
  Serial.begin(115200);
  mouse.begin();
  attachInterrupt(digitalPinToInterrupt(SCL), handleInterrupt, FALLING);

}

void handleInterrupt() {
  noInterrupts();
    for(int i = 0; i < 6; i++){
      mouse.raw[i] = mouse.get(true);
    }
  interrupts();
}

void loop() {
/*
    if(mouse.raw[0] != 0 && mouse.raw[0] != 0x80){
      for (uint8_t i =0; i < 6; i++){
        Serial.print(mouse.raw[i], BIN);
        Serial.print("\t");
      }
      Serial.println();
    }
*/

  if(mouse.raw[0] != 0 && mouse.raw[0] != 0x80){
        Serial.print("Z pressure ");
        Serial.print(mouse.raw[2]);
        Serial.print("\t");
        Serial.print("X position ");
        Serial.print(mouse.raw[4]);
        Serial.print("\t");
        Serial.print("Y position ");
        Serial.println(mouse.raw[5]);
    }
}


example/rawData.cpp
#include <ESP8266WiFi.h>
#include <PS2Mouse.h>

PS2Mouse mouse(SCL, SDA);
void IRAM_ATTR handleInterrupt();

byte raw[6];
void setup() {
  Serial.begin(115200);
  mouse.begin();
  attachInterrupt(digitalPinToInterrupt(SCL), handleInterrupt, FALLING);

}

void handleInterrupt() {
  noInterrupts();
    for(int i = 0; i < 6; i++){
      raw[i] = mouse.get(true);
    }
  interrupts();
}

void loop() {
      if(raw[0] != 0 && raw[0] != 0x80){
        Serial.print(raw[0],BIN);
        Serial.print("\t");
        Serial.print(raw[1],BIN);
        Serial.print("\t");
        Serial.println(raw[2],BIN);
        Serial.print("\t");   
        Serial.print(raw[3],BIN);
        Serial.print("\t");
        Serial.print(raw[4],BIN);
        Serial.print("\t");
        Serial.println(raw[5],BIN);
     }
}


src/PS2Mouse.cpp
#include "PS2Mouse.h"

/*
• $FF Reset
• $FE Resend
• $FC Error
• $FA Acknowledge
• $F6 Set Defaults
• $F5 Disable
• $F4 Enable
• $F3 Set Sample Rate (argument $00-$FF)
• $F2 Read Device Type
• $F0 Set Remote Mode
• $EE Set Wrap Mode
• $EC Reset Wrap Mode
• $EB Read Data
• $EA Set Stream Mode
• $E9 Status Request
• $E8 Set Resolution (argument 0-3)
• $E7 Set Scaling 2:1
• $E6 Set Scaling 1:1
*/

/*
 The Synaptics TouchPad has a small set of configurable features that are encapsulated in the TouchPad mode byte,
 an 8-bit field which the host can set to any value using a special command.

    0x00 Always OK Relative mode                                     // 0xE8 0x00 0xE8 0x00 0xE8 0x00 0xE8 0x00 0xF3 0x14
    0x04 Always OK Relative mode with gestures disabled              // 0xE8 0x00 0xE8 0x00 0xE8 0x01 0xE8 0x00 0xF3 0x14
    0x40 Always OK Relative mode with high packet rate               // 0xE8 0x01 0xE8 0x00 0xE8 0x00 0xE8 0x00 0xF3 0x14
    0x80 capExtended = 0 Absolute mode                               // 0xE8 0x02 0xE8 0x00 0xE8 0x00 0xE8 0x00 0xF3 0x14
  ✔ 0x81 capExtended = 1 Absolute mode with W                        // 0xE8 0x02 0xE8 0x00 0xE8 0x00 0xE8 0x01 0xF3 0x14
    0xC0 capExtended = 0 Absolute mode with high packet rate         // 0xE8 0x03 0xE8 0x00 0xE8 0x00 0xE8 0x00 0xF3 0x14
    0xC1 capExtended = 1 Absolute mode with W, high packet rate      // 0xE8 0x03 0xE8 0x00 0xE8 0x00 0xE8 0x01 0xF3 0x14
    0x0C capSleep = 1 Low-power sleep mode                           // 0xE8 0x00 0xE8 0x00 0xE8 0x03 0xE8 0x00 0xF3 0x14

  *** the function send_tp_arg is intended for this, but it does not always work.
*/

PS2Mouse::PS2Mouse(uint8_t scl, uint8_t sda){
   _scl = scl;
   _sda = sda;
}

PS2Mouse::PS2Mouse(){
  _scl = D1;
  _sda = D2;
}

bool PS2Mouse::begin(){
  byte minor, major, modelCode;
  for (uint8_t i =0; i < 3; i++){                       // -“how many times have you rebooted your computer?”
    send(0xFF);                                         // -“three dude, you always say 'overload three times'” =)
    delay(1000);                                        // Note: This may need an extra-long timeout.
    if (get() != 0xAA){
      error("init, expected 0xAA");
      return false;
    }
    if (get() != 0x00){
      error("init, expected 0x00");
      return false;
    }
  }
  send_tp_arg(0x00);                                    // Send “Identify TouchPad” sequence
  send(0xE9);                                           // Status Request command
  minor = get();                                        // First status byte: TouchPad minor rev.
  modelCode = get();                                    // Second infoModelCode, status byte: 0x47 == Synaptics TouchPad
  major = get();                                        // Third status byte: Major rev in low 4 bits
  Serial.println();
  Serial.print("TouchPad: ");
  (modelCode == 0x47) ? Serial.print("Synaptics") : Serial.print("unknown");
  Serial.print(", id: ");
  Serial.print(modelCode, BIN);
  Serial.print(", minor rev.: ");
  Serial.print(minor, BIN);
  Serial.print(", major: ");
  Serial.print(major, BIN);

  if(modelCode == 0x47){
    send(0xF5);                                         // Disable before activate Absolute mode
    /*Synaptics mode byte (6 Bit)*/
    //send_tp_arg(0x81);                                //it does not always work I do not know the reason
    send(0xE8); send(0x02); send(0xE8); send(0x00); send(0xE8); send(0x00); send(0xE8); send(0x01); send(0xF3); send(0x14);
  }

  //send(0xE6);                                         // E6 Set Scaling 1:1 / E7 Set Scaling 2:1

  //send(0xE8);                                         // E8 Set Resolution (argument 0-3)
  //send(0x00);                                         // Sample Resolution argument of 3

  //send(0xF3);                                         // F3 Set Sample Rate (argument $00-$FF)
  //send(0x14);                                         // Sample Rate argument Valid sample rates are 10, 20, 40, 60, 80, 100, and 200 samples/sec

  send(0xF4);                                           // Enable command
  set_CLOCK(1);                                         // Release PS/2 bus inhibit
  Serial.println(", state: init completed");
  return true;
}

void PS2Mouse::send(byte value){
  byte i, ack, p = 1;
  set_CLOCK(0);
  delayMicroseconds(100);
  set_DATA(0);
  set_CLOCK(1);
  for (i = 0; i < 8; i++) {
    wait_CLOCK(0);
     set_DATA(value & 0x01);
     p = p + value;
     wait_CLOCK(1);
     value = value >> 1;
   }
  wait_CLOCK(0);
  set_DATA(p & 0x01);
  wait_CLOCK(1);
  wait_CLOCK(0);
  set_DATA(1);
  wait_CLOCK(1);
  wait_CLOCK(0);
  if (read_DATA() == 1){
    error("Missing line control bit!");
  }
  wait_CLOCK(1);
  ack = get();
  if (ack != 0xFA){
    error("Probably got an FE or FC error code");
  }
}

byte PS2Mouse::get(bool handler){
  byte i, bit, value = 0, p = 0;
  set_CLOCK(1);
  wait_CLOCK(0);
  wait_CLOCK(1);
  for (i = 0; i < 8; i++){
    wait_CLOCK(0);
    bit = read_DATA();
    value = value + (bit << i);
    p = p + bit;
    wait_CLOCK(1);
  }
  wait_CLOCK(0);
  p = p + read_DATA();
  if ((p & 0x01) == 0){
    error("Parity!");
  }
  wait_CLOCK(1);
  wait_CLOCK(0);
  if (read_DATA() == 0){
    error("Framing!");
  }
  delayMicroseconds(50);
  if(!handler){
    set_CLOCK(0);
  }
  return value;
}

void PS2Mouse::error(const char* e){
  Serial.print("Error: ");
  Serial.println(e);
}

void PS2Mouse::wait_CLOCK(byte state){
  unsigned long saveTime = millis();
  while (digitalRead(_scl) != state){
    if(millis() - saveTime > 50){
      error("Timeout SCL");
      break;
    }
  }
}

void PS2Mouse::set_CLOCK(byte state){
  state == 1 ? high(_scl) : low(_scl);
}

byte PS2Mouse::read_DATA(){
  return digitalRead(_sda);
}

void PS2Mouse::set_DATA(byte data){
  data == 1 ? high(_sda) : low(_sda);
}

void PS2Mouse::send_tp_arg(byte arg){
    byte i;
    for (i = 0; i < 4; i++){
      send(0xE8);
      send((arg >> (6-2*i)) & 3);
    };
}

void PS2Mouse::high(uint8_t pin) {
    pinMode(pin, INPUT);
    digitalWrite(pin, HIGH);
}

void PS2Mouse::low(uint8_t pin) {
    pinMode(pin, OUTPUT);
    digitalWrite(pin, LOW);
}

void PS2Mouse::getAbsoluteAxis(byte *raw, uint16_t &x, uint16_t &y){
  x = ((raw[3] & 0x10) >> 4) << 12 | (raw[1] & 0xF) << 8 | raw[4];
  y = ((raw[3] & 0x20) >> 5) << 12 | ((raw[1] & 0xF0) << 4) | raw[5];
}
